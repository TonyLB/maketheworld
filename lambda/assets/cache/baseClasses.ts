import { ComponentRenderItem } from "@tonylb/mtw-wml/dist/normalize"

export type EphemeraCondition = {
    dependencies: string[];
    if: string;
}

export type EphemeraFeatureAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
}

export type EphemeraFeature = {
    EphemeraId: string;
    key: string;
    tag: 'Feature';
    appearances: EphemeraFeatureAppearance[];
}

export type EphemeraExit = {
    name: string;
    to: string;
}

export type EphemeraRoomAppearance = {
    conditions: EphemeraCondition[];
    name: string;
    render: ComponentRenderItem[];
    exits: EphemeraExit[];
}

export type EphemeraRoom = {
    EphemeraId: string;
    key: string;
    tag: 'Room';
    appearances: EphemeraRoomAppearance[];
}

export type EphemeraMapRoom = {
    EphemeraId: string;
    x: number;
    y: number;
}

export type EphemeraMapAppearance = {
    conditions: EphemeraCondition[];
    fileURL: string;
    name: string;
    rooms: Record<string, EphemeraMapRoom>;
}

export type EphemeraMap = {
    EphemeraId: string;
    key: string;
    tag: 'Map';
    appearance: EphemeraMapAppearance;
}

//
// TODO (ISS1385): Add EphemeraAction to base classes
//

export type EphemeraItem = EphemeraFeature | EphemeraRoom | EphemeraMap
