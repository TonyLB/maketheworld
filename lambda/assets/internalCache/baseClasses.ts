export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
