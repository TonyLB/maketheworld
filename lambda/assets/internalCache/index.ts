import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'
import { delayPromise } from '@tonylb/mtw-utilities/dist/dynamoDB/delayPromise'

export class CacheBase {
    async get(props: { category: never; key: never }) {
        return undefined
    }
    async clear() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

type CacheConstructor = Constructor<{
    get(props: { category: never; key: never; }): Promise<undefined>;
    clear(): void;
}>

type CacheGlobalData = {
    connectionId: string;
    RequestId: string;
}

export const CacheGlobal = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Global: Partial<CacheGlobalData> = {}

        override get(props: { category: 'Global'; key: keyof CacheGlobalData}): Promise<CacheGlobalData[typeof props["key"]]>
        override get(...args: Parameters<InstanceType<GBase>["get"]>): ReturnType<InstanceType<GBase>["get"]>
        override async get(props: any) {
            if (props.category === 'Global') {
                return this.Global[props.key]
            }
            return await super.get(props)
        }

        override clear() {
            this.Global = {}
            super.clear()
        }

        set(props: { category: 'Global', key: 'connectionId' | 'RequestId', value: string }): void
        set({ key, value }: { category: 'Global', key: keyof CacheGlobalData, value: any }): void {
            this.Global[key] = value
        }
    }
}

type CacheLookupOnceData = {
    player: string;
}

export const CacheLookupOnce = <GBase extends Constructor<{ get(props: { category: 'Global', key: 'connectionId' }): Promise<string>; clear(): void; }>>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Lookup: CacheLookupOnceData | undefined = undefined

        override get(props: { category: 'Lookup'; key: 'player'}): Promise<string>
        override get(...args: Parameters<InstanceType<GBase>["get"]>): ReturnType<InstanceType<GBase>["get"]>
        override async get(props: any) {
            if (props.category === 'Lookup') {
                if (this.Lookup === undefined) {
                    const connectionId = await super.get({
                        category: 'Global',
                        key: 'connectionId'
                    })
                    if (connectionId) {
                        //
                        // TODO: Replace repeated attempts with exponential backoff by
                        // refactoring connectionDB.getItem to allow a consistent argument
                        // that can actviate strongly-consistent reads
                        //
                        let attempts = 0
                        let exponentialBackoff = 50
                        while(attempts < 5) {
                            const { player = '' } = await connectionDB.getItem<{ player: string }>({
                                ConnectionId: connectionId,
                                DataCategory: 'Meta::Connection',
                                ProjectionFields: ['player']
                            }) || {}
                            if (player) {
                                return { player }
                            }
                            attempts += 1
                            await delayPromise(exponentialBackoff)
                            exponentialBackoff = exponentialBackoff * 2
                        }
                    }
                }
                return this.Lookup?.[props.key]
            }
            return await super.get(props)
        }

        override clear() {
            this.Lookup = undefined
            super.clear()
        }

    }
}

const InternalCache = CacheLookupOnce(CacheGlobal(CacheBase))
export const internalCache = new InternalCache()
export default internalCache
