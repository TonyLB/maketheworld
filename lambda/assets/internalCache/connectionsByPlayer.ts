import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export class CacheConnectionsByPlayerData {
    ConnectionsByPlayer?: Record<string, string[]>;
    clear() {
        this.ConnectionsByPlayer = undefined
    }
    async get(player: string): Promise<string[]> {
        if (!this.ConnectionsByPlayer) {
            const { connections = {} } = (await connectionDB.getItem<{ connections: Record<string, string> }>({
                ConnectionId: 'Global',
                DataCategory: 'Connections',
                ProjectionFields: ['connections']
            })) || {}
            this.ConnectionsByPlayer = Object.entries(connections)
                .reduce<Record<string, string[]>>((previous, [key, value]) => ({
                    ...previous,
                    [value]: [
                        ...(previous[value] || []),
                        key
                    ]
                }), {})
        }
        return (this.ConnectionsByPlayer || {})[player] || []
    }
}

export const CacheConnectionsByPlayer = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheConnectionsByPlayer extends Base {
        ConnectionsByPlayer: CacheConnectionsByPlayerData = new CacheConnectionsByPlayerData()

        override clear() {
            this.ConnectionsByPlayer.clear()
            super.clear()
        }
    }
}

export default CacheConnectionsByPlayer
