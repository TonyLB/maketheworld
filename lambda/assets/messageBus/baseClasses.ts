import { InternalMessageBus } from '@tonylb/mtw-internal-bus/dist'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/dist'
import { AssetPlayerSettingsAPIMessage, ParseWMLAPIImage } from '@tonylb/mtw-interfaces/dist/asset';
import { EphemeraAssetId, EphemeraCharacterId } from '@tonylb/mtw-interfaces/dist/baseClasses';
import { LibraryAsset, LibraryCharacter } from '@tonylb/mtw-interfaces/dist/library';

export type ReturnValueMessage = {
    type: 'ReturnValue';
    body: Record<string, any>;
}

export type FetchLibraryMessage = {
    type: 'FetchLibrary';
}

type ImportFromAssetArgument = {
    assetId: `ASSET#${string}`;
    keys: string[];
}

export type FetchImportsMessage = {
    type: 'FetchImports';
    importsFromAsset: ImportFromAssetArgument[];
}

export type FetchAssetMessage = {
    type: 'FetchAsset';
    AssetId?: string;
    fileName?: string;
}

type UploadURLMessageImage = {
    key: string;
    contentType: string;
}

export type UploadURLMessage = {
    type: 'UploadURL';
    assetType: 'Character' | 'Asset';
    images: UploadURLMessageImage[];
}

//
// TODO: Refactor ParseWMLMessage so that it pulls the AssetId field from the
// upload, rather than requiring a specific AssetWorkspaceAddress.  Add a "create"
// flag to explicitly authorize creation of a new asset file (rather than update),
// and otherwise fail if the Asset doesn't exist.  Fail with access rejection if
// the asset exists but the player is not authorized to update.
//
export type ParseWMLMessage = {
    type: 'ParseWML';
    AssetId: EphemeraCharacterId | EphemeraAssetId;
    uploadName: string;
    images?: ParseWMLAPIImage[];
    create?: boolean;
}

export type FormatImageMessage = {
    type: 'FormatImage';
    fileName: string;
    width: number;
    height: number;
    AssetId: string;
    imageKey: string;
}

export type MoveAssetMessage = {
    type: 'MoveAsset';
    from: AssetWorkspaceAddress;
    to: AssetWorkspaceAddress;
}

export type MoveByAssetIdMessage = {
    type: 'MoveByAssetId',
    AssetId: string;
    toZone: 'Canon' | 'Library' | 'Personal';
    player?: string;
}

export type LibrarySubscribeMessage = {
    type: 'LibrarySubscribe';
}

export type LibraryUnsubscribeMessage = {
    type: 'LibraryUnsubscribe';
}

export type PlayerInfoMessage = {
    type: 'PlayerInfo';
    player?: string;
    RequestId?: string;
}

export type PlayerSettingsMessage = {
    type: 'PlayerSettings';
    player?: string;
    RequestId?: string;
} & Omit<AssetPlayerSettingsAPIMessage, 'message'>

export type LibraryUpdateMessage = {
    type: 'LibraryUpdate';
}

export type RemoveAssetMessage = {
    type: 'RemoveAsset';
    AssetId: string;
}

export type MessageType = ReturnValueMessage |
    FetchLibraryMessage |
    FetchImportsMessage |
    FetchAssetMessage |
    UploadURLMessage |
    FormatImageMessage |
    ParseWMLMessage |
    MoveAssetMessage |
    MoveByAssetIdMessage |
    LibrarySubscribeMessage |
    LibraryUnsubscribeMessage |
    PlayerInfoMessage |
    PlayerSettingsMessage |
    LibraryUpdateMessage |
    RemoveAssetMessage

export const isReturnValueMessage = (prop: MessageType): prop is ReturnValueMessage => (prop.type === 'ReturnValue')
export const isFetchLibraryAPIMessage = (prop: MessageType): prop is FetchLibraryMessage => (prop.type === 'FetchLibrary')
export const isFetchAssetAPIMessage = (prop: MessageType): prop is FetchAssetMessage => (prop.type === 'FetchAsset')
export const isFetchImportsAPIMessage = (prop: MessageType): prop is FetchImportsMessage => (prop.type === 'FetchImports')
export const isUploadURLMessage = (prop: MessageType): prop is UploadURLMessage => (prop.type === 'UploadURL')
export const isFormatImageMessage = (prop: MessageType): prop is FormatImageMessage => (prop.type === 'FormatImage')
export const isParseWMLMessage = (prop: MessageType): prop is ParseWMLMessage => (prop.type === 'ParseWML')
export const isMoveAssetMessage = (prop: MessageType): prop is MoveAssetMessage => (prop.type === 'MoveAsset')
export const isMoveByAssetIdMessage = (prop: MessageType): prop is MoveByAssetIdMessage => (prop.type === 'MoveByAssetId')
export const isLibrarySubscribeMessage = (prop: MessageType): prop is LibrarySubscribeMessage => (prop.type === 'LibrarySubscribe')
export const isLibraryUnsubscribeMessage = (prop: MessageType): prop is LibraryUnsubscribeMessage => (prop.type === 'LibraryUnsubscribe')
export const isPlayerInfoMessage = (prop: MessageType): prop is PlayerInfoMessage => (prop.type === 'PlayerInfo')
export const isPlayerSettingMessage = (prop: MessageType): prop is PlayerSettingsMessage => (prop.type === 'PlayerSettings')
export const isLibraryUpdateMessage = (prop: MessageType): prop is LibraryUpdateMessage => (prop.type === 'LibraryUpdate')
export const isRemoveAssetMessage = (prop: MessageType): prop is RemoveAssetMessage => (prop.type === 'RemoveAsset')

export class MessageBus extends InternalMessageBus<MessageType> {}
