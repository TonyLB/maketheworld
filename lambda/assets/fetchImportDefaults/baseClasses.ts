import { EphemeraAssetId } from "@tonylb/mtw-interfaces/ts/baseClasses"
import { JSONFileCache } from "../internalCache/jsonFile"
import { Graph } from "@tonylb/mtw-utilities/dist/graphStorage/utils/graph"
import { AssetWorkspaceAddress } from "@tonylb/mtw-asset-workspace"
import ReadOnlyAssetWorkspace from "@tonylb/mtw-asset-workspace/dist/readOnly"

export class InheritanceGraph extends Graph<EphemeraAssetId, { key: EphemeraAssetId; address: AssetWorkspaceAddress }, {}> {}

export class FetchImportsJSONHelper {
    _inheritanceGraph: InheritanceGraph
    constructor(inheritanceGraph: InheritanceGraph) {
        this._inheritanceGraph = inheritanceGraph
    }

    async get(assetId: EphemeraAssetId): Promise<JSONFileCache> {
        const node = this._inheritanceGraph.nodes[assetId]
        if (node) {
            const assetWorkspace = new ReadOnlyAssetWorkspace(node.address)
            await assetWorkspace.loadJSON()
            return {
                normal: assetWorkspace.normal || {},
                namespaceIdToDB: assetWorkspace.namespaceIdToDB || {}
            }    
        }
        else {
            return {
                normal: {},
                namespaceIdToDB: {}
            }
        }
    }
}

