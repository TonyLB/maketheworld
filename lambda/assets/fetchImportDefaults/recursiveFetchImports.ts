import { AssetKey, splitType } from "@tonylb/mtw-utilities/dist/types";
import { isNormalImport, NormalImport } from "@tonylb/mtw-wml/dist/normalize/baseClasses";
import { isSchemaExit } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { isSchemaRoomContents } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { isSchemaRoom } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import { SchemaTag } from "@tonylb/mtw-wml/dist/schema/baseClasses";
import internalCache from "../internalCache";
import { objectMap } from "../lib/objects";
import normalSubset from "./normalSubset"

//
// The translateToFinal class accepts:
//    (a) a current set of occupied keys, and
//    (b) a localToFinal mapping
// and can:
//    (a) Translate SchemaTags to their final key encoding
//    (b) Create new, non-colliding, stub-keys from local keys
//    (c) Push a new mapping that multiplies the existing localToFinal mapping by an importMapping
//    (d) Pop a mapping
//
type NestedTranslateLocalToFinal = Record<string, string>
export class NestedTranslateImportToFinal extends Object {
    localToFinal: NestedTranslateLocalToFinal
    localKeys: string[]
    localStubKeys: string[]
    occupiedFinalKeys: string[]
    constructor(localKeys: string[], localStubKeys: string[], occupiedFinalKeys?: string[], localToFinal?: NestedTranslateLocalToFinal) {
        super()
        if (occupiedFinalKeys) {
            this.occupiedFinalKeys = occupiedFinalKeys
        }
        else {
            this.occupiedFinalKeys = [...localKeys, ...localStubKeys]
        }
        this.localKeys = localKeys
        this.localStubKeys = localStubKeys
        if (localToFinal) {
            this.localToFinal = localToFinal
        }
        else {
            this.localToFinal = [...localKeys, ...localStubKeys].reduce<Record<string, string>>((previous, key) => ({ ...previous, [key]: key }), {})
        }
    }
    nestMapping(keys: string[], stubKeys: string[], mapping: NormalImport["mapping"]): NestedTranslateImportToFinal {
        const localToImport = Object.assign({}, ...Object.entries(mapping).map(([from, { key }]) => ({ [from]: key }))) as Record<string, string>
        const keyMapping = keys
            .reduce<NestedTranslateLocalToFinal>((previous, key) => {
                if (key in localToImport) {
                    return {
                        ...previous,
                        [localToImport[key]]: this.localToFinal[key]
                    }
                }
                return previous
            }, {})
        const stubKeyMapping = stubKeys
            .reduce<NestedTranslateLocalToFinal>((previous, key) => {
                if (key in localToImport) {
                    return {
                        ...previous,
                        [localToImport[key]]: this.localToFinal[key]
                    }
                }
                return previous
            }, {})
        const newTranslate = new NestedTranslateImportToFinal(Object.keys(keyMapping), Object.keys(stubKeyMapping), this.occupiedFinalKeys, { ...stubKeyMapping, ...keyMapping })
        return newTranslate
    }
    translateKey(key: string): string {
        if (key in this.localToFinal) {
            return this.localToFinal[key]
        }
        //
        // TODO: Create collision-detection and avoidance
        //
        return key
    }
    addTranslation(key: string, final: string): void {
        this.localToFinal[key] = final
    }
    translateSchemaTag(tag: SchemaTag): SchemaTag {
        if (isSchemaExit(tag)) {
            return {
                ...tag,
                key: `${this.translateKey(tag.from)}#${this.translateKey(tag.to)}`,
                to: this.translateKey(tag.to),
                from: this.translateKey(tag.from)
            }
        }
        if (isSchemaRoom(tag)) {
            return {
                ...tag,
                key: this.translateKey(tag.key),
                contents: tag.contents.map((value) => (this.translateSchemaTag(value))).filter(isSchemaRoomContents)
            }
        }
        return tag
    }
}

type RecursiveFetchImportArgument = {
    assetId: `ASSET#${string}`;
    translate: NestedTranslateImportToFinal;
    prefixStubKeys?: boolean;
}

export const recursiveFetchImports = async ({ assetId, translate, prefixStubKeys }: RecursiveFetchImportArgument): Promise<SchemaTag[]> => {
    const { localKeys: keys, localStubKeys: stubKeys } = translate
    const { normal } = await internalCache.JSONFile.get(assetId)
    //
    // Coming straight from the datalake, this normal should already be in standardized form,
    // and can be fed directly to normalSubset
    //
    const { newStubKeys, schema } = normalSubset({ normal, keys, stubKeys })
    newStubKeys.forEach((key) => {
        translate.addTranslation(key, prefixStubKeys ? `${splitType(assetId)[1]}.${key}` : key)
    })

    const relevantImports = Object.values(normal)
        .filter(isNormalImport)
        .reduce<{ assetId: `ASSET#${string}`; mapping: NormalImport["mapping"] }[]>((previous, { from, mapping }) => {
            if ([...keys, ...stubKeys, ...newStubKeys].find((key) => (key in mapping))) {
                return [
                    ...previous,
                    {
                        assetId: `ASSET#${from}`,
                        mapping
                    }
                ]
            }
            else {
                return previous
            }
        }, [])
    //
    // Use nested translators from relevantImports to generate recursiveKeyFetches, and then to map
    // the return values to local names
    //
    const importSchema = (await Promise.all(relevantImports.map(async ({ assetId, mapping }) => {
        const nestedTranslate = translate.nestMapping(keys, [...stubKeys, ...newStubKeys], mapping)
        const tags: SchemaTag[] = await recursiveFetchImports({ assetId, translate: nestedTranslate, prefixStubKeys: true })
        return tags.map((tag) => (nestedTranslate.translateSchemaTag(tag)))
    }))).flat()

    return [
        ...importSchema,
        ...schema.map((tag) => (translate.translateSchemaTag(tag)))
    ]

}

export default recursiveFetchImports
