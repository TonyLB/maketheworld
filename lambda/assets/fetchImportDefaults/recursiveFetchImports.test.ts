import Normalizer from '@tonylb/mtw-wml/dist/normalize'
import recursiveFetchImports, { NestedTranslateImportToFinal } from './recursiveFetchImports'

jest.mock('../internalCache')
import internalCache from '../internalCache'
import { NormalForm } from '@tonylb/mtw-wml/dist/normalize/baseClasses'

const internalCacheMock = jest.mocked(internalCache, true)

const testNormalFromWML = (wml: string): NormalForm => {
    const normalizer = new Normalizer()
    normalizer.loadWML(wml)
    return normalizer.normal

}

describe('recursiveFetchImports', () => {
    const testFinal = testNormalFromWML(`<Asset key=(testFinal)>
        <Room key=(testNonImport)>
            <Description>
                DescriptionOne
            </Description>
            <Exit to=(testNonImportStub)>test exit</Exit>
        </Room>
        <Room key=(testNonImportStub)>
            <Name>StubOne</Name>
        </Room>
        <Room key=(testImportOne)>
            <Description>
                Two
            </Description>
            <Exit to=(testImportStubOne)>test exit one</Exit>
        </Room>
        <Room key=(testImportStubOne) />
        <Import from=(testImportAssetOne)>
            <Use type="Room" key=(testImportOne) />
            <Use type="Room" key=(testImportStubOne) />
            <Use type="Room" key=(testImportFoo) as=(testImportTwo) />
        </Import>
    </Asset>`)
    const testImportOne = testNormalFromWML(`<Asset key=(testImportAssetOne)>
        <Room key=(testImportOne)>
            <Description>
                One
            </Description>
        </Room>
        <Room key=(testImportStubOne)>
            <Name>StubTwo</Name>
        </Room>
        <Room key=(testImportFoo)>
            <Description>
                Foo
            </Description>
        </Room>
    </Asset>`)
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        internalCacheMock.JSONFile.get.mockImplementation(async (assetId: string) => {
            let normal: NormalForm = {}
            switch(assetId) {
                case 'ASSET#testFinal':
                    normal = testFinal
                    break
                case 'ASSET#testImportAssetOne':
                    normal = testImportOne
                    break
            }
            return {
                normal,
                namespaceIdToDB: {}
            }
        })
    })

    it('should return empty when passed no keys', async () => {
        expect(await recursiveFetchImports({ assetId: 'ASSET#testFinal', translate: new NestedTranslateImportToFinal([], []) })).toEqual([])
    })

    it('should return element and stubs when passed non-import key', async () => {
        expect(await recursiveFetchImports({ assetId: 'ASSET#testFinal', translate: new NestedTranslateImportToFinal(['testNonImport'], []) })).toEqual([{
            tag: 'Room',
            key: 'testNonImport',
            name: [],
            render: [{ tag: 'String', value: 'DescriptionOne' }],
            contents: [{
                key: 'testNonImport#testFinal.testNonImportStub',
                tag: 'Exit',
                name: 'test exit',
                from: 'testNonImport',
                to: 'testFinal.testNonImportStub',
                contents: []
            }]
        },
        {
            tag: 'Room',
            key: 'testFinal.testNonImportStub',
            name: [{ tag: 'String', value: 'StubOne' }],
            render: [],
            contents: []
        }])
    })

    it('should recursive fetch one level of element and stubs when passed import key', async () => {
        expect(await recursiveFetchImports({ assetId: 'ASSET#testFinal', translate: new NestedTranslateImportToFinal(['testImportOne'], []) })).toEqual([{
            tag: 'Room',
            key: 'testImportOne',
            name: [],
            render: [{ tag: 'String', value: 'One' }],
            contents: []
        },
        {
            tag: 'Room',
            key: 'testFinal.testImportStubOne',
            name: [{ tag: 'String', value: 'StubTwo' }],
            render: [],
            contents: []
        },
        {
            tag: 'Room',
            key: 'testImportOne',
            name: [],
            render: [{ tag: 'String', value: 'Two' }],
            contents: [{
                key: 'testImportOne#testFinal.testImportStubOne',
                tag: 'Exit',
                name: 'test exit one',
                from: 'testImportOne',
                to: 'testFinal.testImportStubOne',
                contents: []
            }]
        },
        {
            tag: 'Room',
            key: 'testFinal.testImportStubOne',
            name: [],
            render: [],
            contents: []
        }])
    })

})