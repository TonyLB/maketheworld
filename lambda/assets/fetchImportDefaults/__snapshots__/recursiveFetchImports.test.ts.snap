// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursiveFetchImports should follow dynamic renames in imports 1`] = `
Array [
  Object {
    "contents": Array [],
    "key": "testImportTwo",
    "name": Array [
      Object {
        "tag": "String",
        "value": "StubFoo",
      },
    ],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "from": "testNonImportTwo",
        "key": "testNonImportTwo#testImportTwo",
        "name": "test exit",
        "tag": "Exit",
        "to": "testImportTwo",
      },
    ],
    "key": "testNonImportTwo",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportTwo",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
]
`;

exports[`recursiveFetchImports should import multilevel and avoid colliding stub names 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "from": "testImportThree",
        "key": "testImportThree#testImportAssetThree.stub",
        "name": "test exit",
        "tag": "Exit",
        "to": "testImportAssetThree.stub",
      },
    ],
    "key": "testImportThree",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportAssetTwo.stub",
    "name": Array [
      Object {
        "tag": "String",
        "value": "Asset Three",
      },
    ],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportAssetThree.stub",
    "name": Array [
      Object {
        "tag": "String",
        "value": "AssetThreeStub",
      },
    ],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "from": "testImportThree",
        "key": "testImportThree#testImportAssetTwo.stub",
        "name": "test exit",
        "tag": "Exit",
        "to": "testImportAssetTwo.stub",
      },
    ],
    "key": "testImportThree",
    "name": Array [],
    "render": Array [
      Object {
        "tag": "String",
        "value": "Asset Two",
      },
    ],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportAssetTwo.stub",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportThree",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
]
`;

exports[`recursiveFetchImports should properly stub out features in room description 1`] = `
Array [
  Object {
    "contents": Array [],
    "key": "testRoomWithFeatures",
    "name": Array [],
    "render": Array [
      Object {
        "tag": "Link",
        "text": "Test",
        "to": "testFeature",
      },
    ],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testFeature",
    "name": Array [],
    "render": Array [],
    "tag": "Feature",
  },
  Object {
    "contents": Array [],
    "key": "testRoomWithFeatures",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
]
`;

exports[`recursiveFetchImports should recursive fetch one level of element and stubs when passed import key 1`] = `
Array [
  Object {
    "contents": Array [],
    "key": "testImportOne",
    "name": Array [],
    "render": Array [
      Object {
        "tag": "String",
        "value": "One",
      },
    ],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportStubOne",
    "name": Array [
      Object {
        "tag": "String",
        "value": "StubTwo",
      },
    ],
    "render": Array [],
    "tag": "Room",
  },
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "from": "testImportOne",
        "key": "testImportOne#testImportStubOne",
        "name": "test exit one",
        "tag": "Exit",
        "to": "testImportStubOne",
      },
    ],
    "key": "testImportOne",
    "name": Array [],
    "render": Array [
      Object {
        "tag": "String",
        "value": "Two",
      },
    ],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testImportStubOne",
    "name": Array [],
    "render": Array [],
    "tag": "Room",
  },
]
`;

exports[`recursiveFetchImports should return element and stubs when passed non-import key 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [],
        "from": "testNonImport",
        "key": "testNonImport#testNonImportStub",
        "name": "test exit",
        "tag": "Exit",
        "to": "testNonImportStub",
      },
    ],
    "key": "testNonImport",
    "name": Array [],
    "render": Array [
      Object {
        "tag": "String",
        "value": "DescriptionOne",
      },
    ],
    "tag": "Room",
  },
  Object {
    "contents": Array [],
    "key": "testNonImportStub",
    "name": Array [
      Object {
        "tag": "String",
        "value": "StubOne",
      },
    ],
    "render": Array [],
    "tag": "Room",
  },
]
`;
